### ISSUE TITLE
Refactor: Moving from Scikit-Learn to Pytorch

### ISSUE THREAD

**Issue by EdwinB12** [alan-turing-institute/autoemulate#315]
This issue is the top level tracking issue for all tasks for converting the backend of Scikit-Learn to Pytorch. 

## Motivation

- GPU Acceleration
- ???
-------------------------------

## What is currently implemented?
-------------------------------
## Why Pytorch?
We have chosen Pytorch over Jax due to the following reasons: 

- More familiarity within the team with Pytorch over Jax
- A more mature ecosystem for providing support for multiple model types - particularly around Guassian Processes
-------------------------------
## Merging into Autoemulate

- The plan is to merge all refactor changes onto a `dev` branch. This `dev` branch will then be merged into `main` when we are ready. This is done so the AutoEmulate API doesn't become confused. This issue will be closed when `dev` is merged into `main`. 
- Practical tasks will be assigned as sub issues on this issue. All tasks will have the `refactor` label. 
-------------------------------

### 1. Add a feature that allows to extract underlying PyTorch models as standalone emulators

- Essentially keep things as they are, but for PyTorch based emulators have an option to extract them from the sklearn pipeline for use in downstream tasks
- Users interested in this functionality could restrict themselves to those models from the outset (and maybe we can have some kind of exclude-non-pytorch-models style arg to never even fit those)
- This would lose scikit-learn's pipeline functionality which allows to attach PCA/data-standardisation. But we can just add this using nn.Sequential to the actual model architecture (not sure about PCA but definitely z-transforms)
- So the exclude-non-pytorch-models would remove non-pytorch models + data standardisation, and potentially add standardisation to the PyTorch architecture.
- In this scenario, we could remove cross-validation and focus on early stopping to improve speed.

### 2. Re-write AutoEmulate to only use pytorch

 - This would mean losing the scikit-learn ecosystem, including pipelines which allow data processing, cross validation, hyperparameter optimisation, parallelisation, and all non neural network models. 
 - some of these features might be replaceable.... 
 - pipelines are partially replacable with additions to the network architecture (e.g. adding a data standardisation module using `nn.Sequential`)
      - cross-validation can probably be removed and instead we focus on things like early stopping and just do one split? (not sure whether this is best practice)
      - hyperparameter search / parallelisation could be done with [Raytune](https://docs.ray.io/en/latest/tune/index.html) or [optuna](https://github.com/optuna/optuna)
- We'll lose models or have to reimplement them as there does not seem to be a reliable way to convert sklearn objects into torch/jax
      - we'll have GPs, Neural Processes, and MLPs without making any changes
      - radial basis functions should be re-implementable in PyTorch without much effort
           - is reimplementation a good use of our time?
           - see for example this [repo](https://github.com/ArmanMaesumi/torchrbf/tree/master?tab=readme-ov-file#readme)
      - tree based methods need some discussion
           - gradient boosted trees (LightGBM particularly) are seen as a good default algorithm for large datasets where GPs struggle so it would be good to keep LightGBM specifically but we probably do not need the other tree based methods (currently have sklearn implementation of random forrest and gradient boosting)
           - LightGBM has GPU support, is there a viable version of autoemulate where we drop the sklearn tree implementations and just keep LightGBM as is? 
           - What about [pytorch-frame](https://github.com/pyg-team/pytorch-frame/tree/master) like approach, which wraps LightGBM (and XGBoost) inside Pytorch to do hyperparam optimization. Is this enough?
       - SVMs would probably go
          - is this a problem? are they used much?
    - advantages:
        - easier to maintain than models from lots of different packages
        - easier to integrate running on GPU's
        - Raytune might make it easier to to architecture search
        - less messy than supporting both pytorch and scikit-learn in autoemulate
    - main question:
        - do we want non-neural network ML emulators? From some experience, people are indeed using XGBoost and similar models to emulate simulations with larger datasets.

**Comment by sgreenbury** [comment]
Following on from discussion yesterday, a further option for taking this forward:
- For sklearn models that are not straightforwardly implmented in torch (e.g. RandomForest) - initially consider retaining functionality as an sklearn model through a distinct `SklearnBackend` wrapper class (see diagram in [comment](https://github.com/alan-turing-institute/autoemulate/issues/322#issuecomment-2715343042)).
- Over time this backend could be deprecated in favour of a complete torch implementation of these models
